man/pexec.1
.TH Scalable\ Unix\ Tools 1 "24/03/2003" " " ""
.SH NAME
.B pexec 
\- execute any command on selected node.

.B pls
\- list the directory contents

.B pcp
\- copy the file locally

.B pmv
\- move the file locally

.B prm
\- remove the file locally

.B pps
\- process list

.B pfind
\- searches for files in directory hierarchy

.B ptest
\- check file types and compare values on the selected nodes

.B pfps
\- searches the process by the process name

.B pkillps
\- send signal to process by name

.B pkillu
\- send signal to process by effective user

.B pdist
\- distribute the file to selected node

.B phost
\- print or count host in hostlist

.SH SYNOPSIS 
.B pexec
[node selecting options] [COMMAND [ARG]...]

.B pls
[node selecting options] [ls options]

.B pcp
[node selecting options] [cp options]

.B pmv
[node selecting options] [mv options]

.B prm
[node selecting options] [rm options]

.B pps
[node selecting options] [ps options]

.B ptest
[node selecting options] <test options>

.B pfind
[ node selecting options] [find options]

.B pfps
[node selecting options] <process name>

.B pkillps
[node selecting options] [-s signal name] <process name>

.B pkillu
[node selecting options] [-s signal name] <user name>

.B pdist
[node selecting options] -[rp] SOURCE... DEST

.B phost
[node selecting options] -[n]

.SH DESCRIPTION 
.PP
Scalable Unix Tools is a set of frequently used UNIX command for cluster. The usage of each command are described briefly below.

.SH PLS,PCP,PMV,PPS,PRM, and PTEST
.PP
Execute the corresponding command on selected nodes. Note that, the result of 
.B ptest
is the and'ed of all returned value of 'test' on all node.

.SH PEXEC
.PP
Execute COMMAND, with the arguments ARG, on the selected nodes.

.SH PKILLPS and PKILLU
.PP
Exec the 'kill' command on selected node base on its process name or the effective user of the process.
.TP
\fB\-s
specify the signal name. The signal name format is the same as kill(1)

.SH PDIST
.PP
Distribute the file to selected node.
.TP
\fB\-r
copy the directories recursively.
.TP
\fB\-p
preserve the file permission.

.SH PHOST
.PP
Print available hostlist, combining with host selecting options, the command can show on/off-line host(s).
.TP
\fB\-n
Print number of host instead of of host name.

.SH HOST SELECTING OPTIONS
.PP
All Scalable Unix Tools accept the same format of node selecting option. Note that, omitting the node selecting option means '-a' (Omitting the options = the first argument passed is not one of the node selecting option). The default operation for node selection is to apply only to the alive node, that is, the command will not be excuted on any down node even though specified. To force doing the command on down node, use either
.B --scu-forceall
option or one of
.B -aa,
.B -ea,
.B -ha
.B -pa,
options.

.PP
The node information is gather from
.B underlying resource monitoring system (such as Ganglia or SCMS)
command. Specify the node outside the scope of
.B underlying resource monitoring
won't have any effect.

.TP
\fB\-a\fR
select all node.
.TP
\fB\-aa\fR
select all node, including off-line nodes.
.TP
\fB\-e\fR
select all node EXCEPT the current node (The node that invoke the command).
.TP
\fB\-ea\fR
select all node EXCEPT the current node (The node that invoke the command), including off-line nodes.
.TP
\fB\-d\fR
select only off-line nodes.
.TP
\fB\-h <node specification>\fR
specify the list of node in argument. The list of node must be in node specificatino format describe in the next section.
.TP
\fB\-ha <node specification>\fR
like -h, but also including off-line nodes.
.TP
\fB\-p <partition>\fR
read the list of node from file <partition>. It is possible to pre-define the partition file by placing the file in directory your_sysconf_dir/scutools/part (eg. /etc/scutools/part). All SUT command read partition from the pre-defined file first, then trying to open it from the current directory. The format of the partition must correspond to the partition file format describe below.
.TP
\fB\-pa <partition>\fR
Like -p, read the list of node from file <partition>, including off-line nodes.

.SH HOST SPECIFICATION FORMAT
.PP
The host specification format is a list of (a little bit extended) regular expression (see regex(7)) separate by comma (,). For example

.nf
	pexec -h compute.*,frontend hostname
.fi

.PP
executes the command 'hostname' on the frontend plus the node which name precede with 'compute'

.PP
Since the regular expression processing is base on the character, there is an extended syntax for the regular expression to support the numbering range using '<' and '>'. For example

.nf
	pexec -h 'compute<20-100>' hostname
.fi

.PP
executes 'hostname' on the node with the name in range of 20-100 (eg. compute20, compute21, compute22, ... , compute100)

.SH PARTITION FILE FORMAT
.PP
Partition file contains a list of node name, one node per line. The leading '#' may be used as the comment. The node name will be interpreted as the regular expression. For example

.nf
	# partition file for myrinet node
	myri-node<1-5>
	myri-frontend
.fi
.PP
has the same meaning as

.nf
	# partition file for myrinet node
	myri-node1
	myri-node2
	myri-node3
	myri-node4
	myri-node5
	myri-frontend
.fi
.PP
Note that, it is possible to use the same file format as MPICH machine file. The trailing :<number of cpu> will be ignored.

.SH Special SUT options
.PP
All commands accept a special option in the form of --sut-option. The list of acceptable options are
.TP
\fB\-\-sut\-forceall\fR
By default, the command will execute only on the on-line node. --sut-forceall force the command to run on both on-line and off-line node.
.TP
\fB\-\-sut\-max\-rshbg <number of concurrent execution>\fR
Every commands will be spawned at 10 hosts simultaneously by default. This greatly speed up the speed of execution, but all the output will come out simultaneously. You can increase and decrease the default number of concurrent execution by using this option. This can als be specified in sce.conf at [scms] with configuration option "max_rshbg"
.SH CONFIGURATION
All commands read the configuration from [scms] section in sce.conf. Below are the list of configurable option for all commands.
.TP
\fBmax_rshbg\fR
This has the same effect as \-\-sut\-max\-rshbg. For example, max_rshbg=20
.TP
\fBsut_hostlen\fR
Change the position of character to trim hostname. The default position is 14. Change this in case of long hostname.

.SH EXAMPLE
.PP
List the content of the file /etc/fstab on all node.

.nf
	pcat -a /etc/fstab
.fi
.PP
Restart the xinetd daemon on all node except the current.

.nf
	pexec -e /sbin/service xinetd restart
.fi
.PP
List the available ethernet interface on some node.

.nf
	pexec -h "myri<1-5>" /sbin/ifconfig -a
.fi
.PP
Distribute the configuration file to all compute node

.nf
	pdist -h "compute.*" /etc/sce/sce.conf /etc/sce/
.fi
.PP
Start/Stop service on all node, even the off-line node.

.nf
	pexec -aa /etc/init.d/snmpd restart
.fi

.SH AUTHOR
.PP
Written by Somsak Sriprayoonsakul <somsaks@gmail.com>, Sugree Pattanapherom 

.SH REPORTING BUGS
.PP
Report bugs to <somsaks@gmail.com>

.SH SEE ALSO
.BR gstat (1)

.SH REVISION 
.PP
$Id: pexec.1 10543 2005-09-20 17:25:43Z ssy $


scutools/pexec.py
"""
Generic SUT interface
    
@author Somsak Sriprayoonsakul
"""

import sys, string, os

import config, node
from error import *

import spawner

class PExec :
    "Generic interface to parallel command"
    HARG_ALL = 1
    HARG_EXCEPT = 2
    HARG_HOST = 3
    HARG_PART = 4
    HARG_DOWN = 5

    def __init__(self, cmd, prog_args) :
        " Initialized internal variable "
        self.cmd = cmd
        self.cmd_args = None
        self.hostarg = (None, None)
        self.hostlist = []
        self.specarg = {}
        self.parseArg(prog_args)
        self.parseHostArg()

    def parseArg(self, prog_args) :
        "Parse generic sut argument"
        
        try :
            # parse special option first
            i = 0
            while i < len(prog_args) :
                if prog_args[i][0:5] == '--scu' :
                    arg = prog_args[i][6:]
                    if arg :
                        arg = string.split(arg, '=', 1)
                        if len(arg) == 1 :
                            self.specarg[arg[0]] = 1
                        else :
                            self.specarg[arg[0]] = arg[1]
                    del prog_args[i]
                else :
                    i = i + 1
            # now host option
            try :
                self.hostarg = (self.HARG_ALL, None)
                if prog_args[0] == '--help' :
                    # this is special case, since --help is the first command which is not very likely to happen
                    if len(prog_args) < 2 :
                        raise InvalArg, 'try man pexec'
                if prog_args[0] == '-a' or\
                    prog_args[0] == '--all' :
                    self.hostarg = (self.HARG_ALL, None)
                    del prog_args[0]
                elif prog_args[0] == '-aa' or \
                    prog_args[0] == '--really-all' :
                    self.hostarg = (self.HARG_ALL, None)
                    self.specarg['forceall'] = 1
                    del prog_args[0]
                elif prog_args[0] == '-d' or \
                    prog_args[0] == '--down' :
                    self.hostarg = (self.HARG_DOWN, None)
                    self.specarg['forceall'] = 1
                    del prog_args[0]
                elif prog_args[0] == '-e' or\
                    prog_args[0] == '--except' :
                    self.hostarg = (self.HARG_EXCEPT, None)
                    del prog_args[0]
                elif prog_args[0] == '-ea' or\
                    prog_args[0] == '--except-all' :
                    self.hostarg = (self.HARG_EXCEPT, None)
                    self.specarg['forceall'] = 1
                    del prog_args[0]
                elif prog_args[0] == '-p' or\
                    prog_args[0] == '--part' :
                    if len(prog_args) < 2 :
                        self.listPart()
                        raise NoError
                    self.hostarg = (self.HARG_PART, prog_args[1])
                    del prog_args[0:2]
                elif prog_args[0] == '-pa' or\
                    prog_args[0] == '--part-all' :
                    if len(prog_args) < 2 :
                        self.listPart()
                        raise NoError
                    self.hostarg = (self.HARG_PART, prog_args[1])
                    self.specarg['forceall'] = 1
                    del prog_args[0:2]
                elif prog_args[0] == '-h' or\
                    prog_args[0] == '--host' :
                    if len(prog_args) < 2 :
                        raise InvalArg, 'need at least one host pattern, such as \'compute.*\''
                    self.hostarg = (self.HARG_HOST, prog_args[1])
                    del prog_args[0:2]
                elif prog_args[0] == '-ha' or\
                    prog_args[0] == '--host-all' :
                    if len(prog_args) < 2 :
                        raise InvalArg, 'need at least one host pattern, such as \'compute<0-9>\''
                    self.hostarg = (self.HARG_HOST, prog_args[1])
                    self.specarg['forceall'] = 1
                    del prog_args[0:2]
            except IndexError :
                pass
                #if not config.compat_scms_2 :
                #    # We can't let the command without host option!
                #    raise InvalArg, 'need host specification'
            # The rest is command option
            self.cmd_args = prog_args

        except IndexError:
            if i < len(prog_args) :
                raise InvalArg, prog_args[0]
            else :
                raise InvalArg, 'none'

    def parseHostArg(self) :
        "Create hostlist from host specification argument"
        # Get node first
        flag = 0x0
        if self.hostarg[0] == self.HARG_EXCEPT :
            flag = flag | node.ALL_EXC

        if self.hostarg[0] == self.HARG_DOWN or self.specarg.has_key('forceall') :
            self.hostlist = node.get_alive(node.ALL | flag)
        else :
            self.hostlist = node.get_alive(node.ALIVE | flag)

        if self.hostarg[0] == self.HARG_DOWN :
            on_hostlist = node.get_alive(node.ALIVE | flag)
            # exclude host that's in on-line hosts
            real_hostlist = []
            for host in self.hostlist :
                if not host in on_hostlist :
                    real_hostlist.append(host)
            self.hostlist = real_hostlist
                    
        # Parse host argument, if needed
        if self.hostarg[0] == self.HARG_HOST or self.hostarg[0] == self.HARG_PART :
            tmp_hostlist = []
            if self.hostarg[0] == self.HARG_PART :
                import partfile
                tmp_hostlist = partfile.part_read(self.hostarg[1])
            elif self.hostarg[0] == self.HARG_HOST :
                tmp_hostlist = string.split(self.hostarg[1], ',')
            # Treat each line as REGEX
            #
            real_hostlist = []
            import xre
            while tmp_hostlist :
                h = tmp_hostlist[0]
                try :
                    ro = xre.compile('^' + h + '$')
                except :
                    if len(tmp_hostlist) != 1 :
                        tmp_hostlist[0] = string.join([h,tmp_hostlist[1]], ',')
                        del tmp_hostlist[1]
                        continue
                    else :
                        raise InvalHost, h
                i = 0
                while i < len(self.hostlist) :
                    # match the whole hostname
                    m1 = ro.match(self.hostlist[i])
                    # additionally match just short hostname
                    m2 = ro.match(self.hostlist[i].split('.')[0])
                    if m1 or m2 :
                        real_hostlist.append(self.hostlist[i])
                        del self.hostlist[i]
                    else :
                        i = i + 1
                del tmp_hostlist[0]
            self.hostlist = real_hostlist
        
        # Move our name to the last of the list, if any
        self.moveLast()
    
    def moveLast(self) :
        "Move hostname to the last entry of list"
        import socket

        myaddr = socket.gethostbyname(socket.gethostname())
        for i in range(len(self.hostlist)) :
            try :
                if socket.gethostbyname(self.hostlist[i]) is myaddr :
                    del self.hostlist[i]
                    self.hostlist.append(h)
                    break
            except socket.gaierror, e :
                raise NodeStatus(self.hostlist[i])

    def launch(self, out = None, err = None) :
        """Execute the command, return the exit status"""
        # Get spawner first
        spawner_arg = config.def_spawner
        if self.specarg.has_key('spawner') :
            spawner_arg = self.specarg['spawner'][1]
        
        launcher = None
        if spawner_arg == 'rsh' :
            launcher = spawner.RshBg(self.specarg)
        else :
            raise InvalLauncher(spawner_arg)

        if not self.cmd and not self.cmd_args :
            raise InvalArg('')

        retval = launcher.spawn(self.hostlist, self.cmd, self.cmd_args, out, err)
        return retval

    def listPart(self) :
        """List partition file available on the system"""
        if config.partfile_dir :
            print 'Part available in this host are:'
            for file in os.listdir(config.partfile_dir) :
                path = os.path.join(config.partfile_dir, file)
                if os.path.isfile(path) :
                    print file
        
if __name__ == '__main__' :
    cmd = PExec(sys.argv[1], sys.argv[2:])
    print 'command = ', cmd.cmd
    print 'command args = ', cmd.cmd_args
    print 'host arg = ', cmd.hostarg
    print 'special args = ', cmd.specarg
    print 'hostlist = ', cmd.hostlist
    cmd.launch()
